File: Source/Zybach.API/src/controllers/manager-dashboard.controller.ts
import { inject } from "inversify";
import { Hidden, Route, Controller, Get, Path, Security } from "tsoa";
import { DistrictStatisticsDto } from "../dtos/district-statistics-dto";
import { streamFlowZonePumpingDepthDto } from "../dtos/stream-flow-zone-pumping-depth-dto";
import { InternalServerError } from "../errors/internal-server-error";
import AghubWell from "../models/aghub-well";
import { RoleEnum } from "../models/role";
import { SecurityType } from "../security/authentication";
import { AghubWellService } from "../services/aghub-well-service";
import { GeoOptixService } from "../services/geooptix-service";
import { InfluxService } from "../services/influx-service";
import GALLON_TO_ACRE_INCH from "../util/constants";
import { provideSingleton } from "../util/provide-singleton";

@Hidden()
@Route("/api/managerDashboard")
@provideSingleton(ManagerDashboardController)
export class ManagerDashboardController extends Controller {
    constructor(
        @inject(GeoOptixService) private geooptixService: GeoOptixService,
        @inject(InfluxService) private influxService: InfluxService,
        @inject(AghubWellService) private aghubWellService: AghubWellService
    ) {
        super();
    }

    @Get("/districtStatistics/{year}")
    @Security(SecurityType.KEYSTONE, [RoleEnum.Adminstrator])
    public async getDistrictStatistics(
        @Path() year: number
    ): Promise<DistrictStatisticsDto> {
        let numberOfWellsTracked, numberOfContinuityMeters, numberOfFlowMeters, numberOfElectricalUsageEstimates;

        // get all wells that either existed in GeoOptix as of the given year or that had electrical estimates as of the given year
        const geoOptixWells = await this.geooptixService.getWellSummariesCreatedAsOfYear(year);
        const aghubRegistrationIds = await this.influxService.getWellRegistrationIdsWithElectricalEstimateAsOfYear(year);
        // combine the registration ids as a set and count to avoid counting duplicates
        numberOfWellsTracked = new Set([...geoOptixWells.map(x => x.wellRegistrationID), ...aghubRegistrationIds]).size;

        // get all sensors that existed in GeoOptix as of the given year
        const sensors = await this.geooptixService.getSensorSummariesCreatedAsOfYear(year);

        // filter by sensor type and count
        numberOfFlowMeters = sensors.filter(x => x.sensorType === 'Flow Meter').length;
        numberOfContinuityMeters = sensors.filter(x => x.sensorType === 'Continuity Meter').length;

        // todo: get total number of electrical usage estimates
        numberOfElectricalUsageEstimates = aghubRegistrationIds.length;

        return {
            NumberOfWellsTracked: numberOfWellsTracked,
            NumberOfContinuityMeters: numberOfContinuityMeters,
            NumberOfElectricalUsageEstimates: numberOfElectricalUsageEstimates,
            NumberOfFlowMeters: numberOfFlowMeters
        }
    }

    @Get("/streamFlowZonePumpingDepths")
    public async getStreamFlowZonePumpingDepths(
    ): Promise<{ year: number, streamFlowZonePumpingDepths: streamFlowZonePumpingDepthDto[] }[]> {
        // Currently, we are only accounting for electrical data when color-coding the SFZ map;
        // hence, we can confine our attention to the aghub wells and the electrical estimate time series

        const currentYear = new Date().getFullYear();
        const years = []

        for (let year = 2019; year <= currentYear; year++) {
            years.push(year);
        }

        // Step 1. Get a mapping from SFZs to Wells
        const streamFlowZoneWellMap = await this.aghubWellService.getAllWellsWithStreamFlowZones();
        const results = await Promise.all(years.map(async year => {
            // Step 2. Get the total pumped volume for each well.
            // This is represented as a mapping from Well Registration IDs to pumped volumes
            let pumpedVolumes: Map<string, number> = await this.influxService.getAnnualEstimatedPumpedVolumeByWellForYear(year);

            // Step 3. For each SFZ, calculate the pumping depth
            const pumpingDepths: streamFlowZonePumpingDepthDto[] = [];
            for (const [zone, wells] of streamFlowZoneWellMap) {
                if (!wells?.length) {
                    pumpingDepths.push({ streamFlowZoneFeatureID: zone.properties.FeatureID, pumpingDepth: 0, totalIrrigatedAcres: 0, totalPumpedVolume:0 });
                    continue;
                }

                // Sum the irrigated acres and pumped volume for each well
                const totals = wells.reduce((runningTotals, currentWell) => {
                    const wellAcres = currentWell.irrigatedAcresPerYear.find(x => x.year === year)?.acres ?? 0
                    const wellPumpedVolume = pumpedVolumes.get(currentWell.wellRegistrationID) ?? 0;

                    return {
                        totalAcres: runningTotals.totalAcres + wellAcres,
                        totalVolume: runningTotals.totalVolume + wellPumpedVolume
                    }
                }, { totalAcres: 0, totalVolume: 0 })

                // todo: this is reporting in gallons/acres right now and we probably want acre-inch per acre
                pumpingDepths.push({ streamFlowZoneFeatureID: zone.properties.FeatureID, pumpingDepth: GALLON_TO_ACRE_INCH * totals.totalVolume / totals.totalAcres, totalIrrigatedAcres: totals.totalAcres, totalPumpedVolume: GALLON_TO_ACRE_INCH * totals.totalVolume })
            }

            return { year, streamFlowZonePumpingDepths: pumpingDepths };
        }));

        return results;
    }
}

File: Source/Zybach.API/Controllers/ManagerDashboardController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Zybach.API.Services;
using Zybach.API.Services.Authorization;
using Zybach.EFModels.Entities;
using Zybach.Models.DataTransferObjects;

namespace Zybach.API.Controllers
{
    [ApiController]
    [ApiExplorerSettings(IgnoreApi = true)]
    public class ManagerDashboardController : SitkaController<ManagerDashboardController>
    {
        private readonly GeoOptixService _geoOptixService;
        private readonly WellService _wellService;
        private const double GALLON_TO_ACRE_INCH = 3.68266E-5;

        public ManagerDashboardController(ZybachDbContext dbContext, ILogger<ManagerDashboardController> logger, KeystoneService keystoneService, IOptions<ZybachConfiguration> zybachConfiguration, GeoOptixService geoOptixService, WellService wellService) : base(dbContext, logger, keystoneService, zybachConfiguration)
        {
            _geoOptixService = geoOptixService;
            _wellService = wellService;
        }


        [HttpGet("/api/managerDashboard/districtStatistics")]
        [ZybachViewFeature]
        public async Task<DistrictStatisticsDto> GetDistrictStatistics()
        {
            var allWells = await _wellService.GetAghubAndGeoOptixWells();

            return new DistrictStatisticsDto
            { 
                NumberOfWellsTracked = allWells.Count(),
                NumberOfContinuityMeters = allWells.Where(x => x.Sensors.Any(y => y.SensorType == InfluxDBService.SensorTypes.ContinuityMeter)).Select(x => x.WellRegistrationID).Distinct().Count(),
                NumberOfElectricalUsageEstimates = allWells.Where(x => x.Sensors.Any(y => y.SensorType == InfluxDBService.SensorTypes.ElectricalUsage)).Select(x => x.WellRegistrationID).Distinct().Count(),
                NumberOfFlowMeters = allWells.Where(x => x.Sensors.Any(y => y.SensorType == InfluxDBService.SensorTypes.FlowMeter)).Select(x => x.WellRegistrationID).Distinct().Count()
            };
        }

        public class DistrictStatisticsDto
        {
            public int NumberOfWellsTracked { get; set; }
            public int NumberOfContinuityMeters { get; set; }
            public int NumberOfElectricalUsageEstimates { get; set; }
            public int NumberOfFlowMeters { get; set; }
        }

        [HttpGet("/api/managerDashboard/streamFlowZonePumpingDepths")]
        public List<AnnualStreamFlowZonePumpingDepthDto> GetStreamFlowZonePumpingDepths()
        {
            // Currently, we are only accounting for electrical data when color-coding the SFZ map;
            // hence, we can confine our attention to the aghub wells and the electrical estimate time series

            var currentYear = DateTime.Today.Year;
            var years = Enumerable.Range(2019, currentYear - 2019 + 1);

            // Step 1. Get a mapping from SFZs to Wells
            var streamFlowZoneWellMap = StreamFlowZone.ListStreamFlowZonesAndWellsWithinZone(_dbContext);
            var annualStreamFlowZonePumpingDepths = new List<AnnualStreamFlowZonePumpingDepthDto>();
            foreach (var year in years)
            {
                annualStreamFlowZonePumpingDepths.Add(CreateAnnualStreamFlowZonePumpingDepthDto(year, streamFlowZoneWellMap));
            }
            return annualStreamFlowZonePumpingDepths;

        }

        private AnnualStreamFlowZonePumpingDepthDto CreateAnnualStreamFlowZonePumpingDepthDto(int year, List<StreamFlowZoneWellsDto> streamFlowZoneWellMap)
        {
            // Step 2. Get the total pumped volume for each well.
            // This is represented as a mapping from Well Registration IDs to pumped volumes
            var pumpedVolumes = WellSensorMeasurement
                .GetWellSensorMeasurementsByMeasurementTypeAndYear(_dbContext, MeasurementTypeEnum.ElectricalUsage, year)
                .GroupBy(x => x.WellRegistrationID).ToDictionary(x => x.Key, x => x.Sum(y => y.MeasurementValue));

            // Step 3. For each StreamFlowZone, calculate the pumping depth
            var streamFlowZonePumpingDepthDtos = new List<StreamFlowZonePumpingDepthDto>();
            foreach (var streamFlowZoneWellsDto in streamFlowZoneWellMap)
            {
                if (!streamFlowZoneWellsDto.Wells.Any())
                {
                    streamFlowZonePumpingDepthDtos.Add(
                        new StreamFlowZonePumpingDepthDto(streamFlowZoneWellsDto.StreamFlowZone.StreamFlowZoneID, 0, 0, 0));
                }
                else
                {
                    var wellRegistrationIDsWithinStreamFlowZone = streamFlowZoneWellsDto.Wells.Select(x => x.WellRegistrationID.ToUpper());
                    var totalIrrigatedAcres = streamFlowZoneWellsDto.Wells
                        .Where(x => wellRegistrationIDsWithinStreamFlowZone.Contains(x.WellRegistrationID))
                        .SelectMany(x => x.IrrigatedAcresPerYear).Where(x => x.Year == year).Sum(x => x.Acres);
                    var totalVolume = pumpedVolumes.Where(x => wellRegistrationIDsWithinStreamFlowZone.Contains(x.Key)).Sum(x => x.Value);

                    // todo: this is reporting in gallons/acres right now and we probably want acre-inch per acre
                    streamFlowZonePumpingDepthDtos.Add(new StreamFlowZonePumpingDepthDto(
                        streamFlowZoneWellsDto.StreamFlowZone.StreamFlowZoneID,
                        GALLON_TO_ACRE_INCH * totalVolume / totalIrrigatedAcres, totalIrrigatedAcres,
                        GALLON_TO_ACRE_INCH * totalVolume));
                }
            }

            return new AnnualStreamFlowZonePumpingDepthDto(year, streamFlowZonePumpingDepthDtos);
        }
    }

    public class AnnualStreamFlowZonePumpingDepthDto
    {
        public AnnualStreamFlowZonePumpingDepthDto()
        {
        }

        public AnnualStreamFlowZonePumpingDepthDto(int year, List<StreamFlowZonePumpingDepthDto> streamFlowZonePumpingDepths)
        {
            Year = year;
            StreamFlowZonePumpingDepths = streamFlowZonePumpingDepths;
        }

        public int Year { get; set; }
        public List<StreamFlowZonePumpingDepthDto> StreamFlowZonePumpingDepths { get; set; }
    }
    public class StreamFlowZonePumpingDepthDto
    {
        public StreamFlowZonePumpingDepthDto()
        {
        }

        public StreamFlowZonePumpingDepthDto(int streamFlowZoneFeatureID, double pumpingDepth, double totalIrrigatedAcres, double totalPumpedVolume)
        {
            StreamFlowZoneFeatureID = streamFlowZoneFeatureID;
            PumpingDepth = pumpingDepth;
            TotalIrrigatedAcres = totalIrrigatedAcres;
            TotalPumpedVolume = totalPumpedVolume;
        }

        public int StreamFlowZoneFeatureID { get; set; }
        public double PumpingDepth { get; set; }
        public double TotalIrrigatedAcres { get; set; }
        public double TotalPumpedVolume { get; set; }

    }
}